pkg/webhook/preflight/generic/registry.go

```diff
@@ -0,0 +1,185 @@
+	if r.registryMirror != nil {
+		return r.checkRegistry(
+			ctx,
+			mirrorURLValidationRegex.ReplaceAllString(r.registryMirror.URL, ""),
+			r.registryMirror.Credentials,
+		)
+	}
+	if r.imageRegistry != nil {
+		return r.checkRegistry(
+			ctx,
+			mirrorURLValidationRegex.ReplaceAllString(r.imageRegistry.URL, ""),
+			r.imageRegistry.Credentials,
+		)
+	}
+	return preflight.CheckResult{
+		Allowed: true,
+	}
+}
+
+func (r *registyCheck) checkRegistry(
+	ctx context.Context,
+	registryURL string,
+	credentials *carenv1.RegistryCredentials,
+) preflight.CheckResult {
+	result := preflight.CheckResult{
+		Allowed: false,
+	}
+	mirrorHost := config.Host{
+		Name: registryURL,
+	}
+	if credentials != nil && credentials.SecretRef != nil {
+		mirrorCredentialsSecret := &corev1.Secret{}
+		err := r.kclient.Get(
+			ctx,
+			types.NamespacedName{
+				Namespace: r.cluster.Namespace,
+				Name:      r.registryMirror.Credentials.SecretRef.Name,
+			},
+			mirrorCredentialsSecret,
+		)
+		if err != nil {
+			result.Allowed = false
+			result.Error = true
+			result.Causes = append(result.Causes,
+				preflight.Cause{
+					Message: fmt.Sprintf("failed to get Registry credentials Secret: %s", err),
+					Field:   fmt.Sprintf("%s.credentials.secretRef", registryMirrorVarPath),
+				},
+			)
+			return result
+		}
+		username, ok := mirrorCredentialsSecret.Data["username"]
+		if !ok {
+			result.Allowed = false
+			result.Error = true
+			result.Causes = append(result.Causes,
+				preflight.Cause{
+					Message: "failed to get username from Registry credentials Secret. secret must have field username.",
+					Field:   fmt.Sprintf("%s.credentials.secretRef", registryMirrorVarPath),
+				},
+			)
+			return result
+		}
+		password, ok := mirrorCredentialsSecret.Data["password"]
+		if !ok {
+			result.Allowed = false
+			result.Error = true
+			result.Causes = append(result.Causes,
+				preflight.Cause{
+					Message: "failed to get username from Registry credentials Secret. secret must have field password.",
+					Field:   fmt.Sprintf("%s.credentials.secretRef", registryMirrorVarPath),
+				},
+			)
+			return result
+		}
+		mirrorHost.User = string(username)
+		mirrorHost.Pass = string(password)
+		if caCert, ok := mirrorCredentialsSecret.Data["ca.crt"]; ok {
+			mirrorHost.RegCert = string(caCert)
+		}
+	}
+	rc := regclient.New(
+		regclient.WithConfigHost(mirrorHost),
+		regclient.WithUserAgent("regclient/example"),
+	)
+	mirrorRef, err := ref.NewHost(mirrorHost.Hostname)
+	if err != nil {
+		result.Allowed = false
+		result.Error = true
+		result.Causes = append(result.Causes,
+			preflight.Cause{
+				Message: "failed to create a client to verify registry configuration",
+				Field:   registryMirrorVarPath,
+			},
+		)
+		return result
+	}
+	_, err = rc.Ping(ctx, mirrorRef) // ping will return an error for anything that's not 200
+	if err != nil {
+		result.Allowed = false
+		result.Error = true
```

Comment:
```suggestion
```

I think the "ping" is the actual check, and its failure is not an internal error.

Reply:
oh this field is only for internal errors? we just need to set the cause instead?

i

Reply:
sounds good i just removed this bit for now.